File: Data Structures/Stack/Java/stack.java
public class Stack<T> {
	
	private T[] array;
	private int top;
	
	public Stack(int size) {
		array = (T[]) new Object[size];
		top = -1;
	}
	
	public boolean isEmpty() {
		return this.top == -1;
	}
	
	public boolean isFull() {
		return this.top == (array.length - 1);
	}
	
	public void push(T element) throws RuntimeException {
		if (isFull())
			throw new RuntimeException();
		
		this.top++;
		this.array[top] = element;
	}
	
	public T pop(T element) throws RuntimeException {
		if (isEmpty())
			throw new RuntimeException();
		
		T removedElement = array[top];
		this.top--;
		
		return removedElement;
	}
	
	public T top() {
		if (isEmpty())
			return null;
		return this.top;
	}



}

File: Data Structures/Stack/JavaScript/stack.js
class Stack {
  constructor () {
    this.items = []
  }
  push (item) {
    this.items.push(item)
  }
  pop () {
    if (!this.items.length) return 'Stack underflow'
    return this.items.pop()
  }
  peek () {
    return this.items[this.items.length - 1]
  }
  isEmpty () {
    return this.items.length ? 1 : 0
  }
  printStack () {
    return this.items.reduce((prev, cur) => prev + ' ' + cur)
  }
}

var stack = new Stack()
stack.push(2)
stack.push(3)
stack.push(4)
console.log(stack.printStack())
console.log(stack.isEmpty())
console.log(stack.pop())
console.log(stack.printStack())
